File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
import collections
import sys
from heapq import heappush,heappop

def solution(n, s, a, b, fares):
    answer=sys.maxsize

    graph=collections.defaultdict(list)

    def dijkstra(start,distance):
        heap=[(0,start)]
        distance[start]=0
        while heap:
            cur_dist,cur_node=heappop(heap)
            if distance[cur_node]<cur_dist:
                continue

            for next_node,next_cost in graph[cur_node]:
                next_dist=distance[cur_node]+next_cost
                if next_dist<distance[next_node]:
                    distance[next_node]=next_dist
                    heappush(heap,(distance[next_node],next_node))

    for u,v,c in fares:
        graph[u].append((v,c))
        graph[v].append((u,c))

    distance_a=[sys.maxsize for _ in range(n+1)]
    distance_b=[sys.maxsize for _ in range(n+1)]
    distance_start=[sys.maxsize for _ in range(n+1)]

    dijkstra(a,distance_a)
    dijkstra(b,distance_b)
    dijkstra(s,distance_start)

    for i in range(1,n+1):
        # if i in (s,a,b):
        #     continue
        answer=min(answer,distance_a[i]+distance_b[i]+distance_start[i])
    return answer

print(solution(6, 4, 6, 2, [[4, 1, 10], [3, 5, 24], [5, 6, 2], [3, 1, 41], [5, 1, 24], [4, 6, 50], [2, 4, 66], [2, 3, 22], [1, 6, 25]]))
print(solution(7, 3, 4, 1, [[5, 7, 9], [4, 6, 4], [3, 6, 1], [3, 2, 3], [2, 1, 6]]))
print(solution(6, 4, 5, 6, [[2, 6, 6], [6, 3, 7], [4, 6, 7], [6, 5, 11], [2, 5, 12], [5, 3, 20], [2, 4, 8], [4, 3, 9]]))

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.cpp
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <map>
#include <queue>
#include <iostream>

using namespace std;
map<int,vector<pair<int,int>>> graph;
int dist[201];

int dijkstra(int start,int end){
    fill(dist,dist+201,INT_MAX);
    
    priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> heap;
    heap.push(make_pair(0,start));
    dist[start]=0;
    
    while(!heap.empty()){
        pair<int,int> p=heap.top();
        heap.pop();
        int curCost=p.first;
        int curNode=p.second;
        
        if (curNode==end){
            return curCost;
        }
        
        if (curCost>dist[curNode]){
            continue;
        }
        
        for (auto adjP:graph[curNode]){
            int adjNode=adjP.first;
            int adjCost=adjP.second;
            
            if (dist[adjNode]>adjCost+curCost){
                dist[adjNode]=adjCost+curCost;
                heap.push(make_pair(dist[adjNode],adjNode));
            }
        }
    }
    return INT_MAX;
}

int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = INT_MAX;
    
    for(auto arr:fares){
        int a =arr[0];
        int b=arr[1];
        int cost=arr[2];
        
        graph[a].push_back(make_pair(b,cost));
        graph[b].push_back(make_pair(a,cost));
        
    }
    
    for (int middleNode=1; middleNode<=n;middleNode++){
        answer=min(answer,dijkstra(s,middleNode)+dijkstra(middleNode,a)+dijkstra(middleNode,b));
    }
    return answer;
}